src/runtime/uv/tcp.cpp

```diff
@@ -0,0 +1,332 @@
+        lean_dec_ref(tcp_socket->m_promise_accept);
+    }
+
+    if(tcp_socket->m_promise_read != NULL) {
+        lean_dec_ref(tcp_socket->m_promise_read);
+    }
+
+    if(tcp_socket->m_byte_array != NULL) {
+        lean_dec_ref(tcp_socket->m_byte_array);
+    }
+
+    event_loop_lock(&global_ev);
+
+    uv_close((uv_handle_t*)tcp_socket->m_uv_tcp, [](uv_handle_t* handle) {
+        free(handle);
+    });
+
+    event_loop_unlock(&global_ev);
+
+    free(tcp_socket);
+}
+
+void initialize_libuv_tcp_socket() {
+    g_uv_tcp_socket_external_class = lean_register_external_class(lean_uv_tcp_socket_finalizer, [](void* obj, lean_object* f) {
+        lean_uv_tcp_socket_object* tcp_socket = (lean_uv_tcp_socket_object*)obj;
+
+        if (tcp_socket->m_promise_accept != NULL) {
+            lean_inc(f);
+            lean_apply_1(f, tcp_socket->m_promise_accept);
+        }
+
+        if (tcp_socket->m_promise_read != NULL) {
+            lean_inc(f);
+            lean_apply_1(f, tcp_socket->m_promise_read);
+        }
+
+        if (tcp_socket->m_byte_array != NULL) {
+            lean_inc(f);
+            lean_apply_1(f, tcp_socket->m_byte_array);
+        }
+    });
+}
+
+// TCP Socket Operations
+extern "C" LEAN_EXPORT lean_obj_res lean_uv_tcp_new() {
+    lean_uv_tcp_socket_object * tcp_socket = (lean_uv_tcp_socket_object*)malloc(sizeof(lean_uv_tcp_socket_object));
+
+    tcp_socket->m_promise_accept = NULL;
+    tcp_socket->m_promise_read = NULL;
+    tcp_socket->m_byte_array = NULL;
+
+    uv_tcp_t * uv_tcp = (uv_tcp_t*)malloc(sizeof(uv_tcp_t));
+
+    event_loop_lock(&global_ev);
+    int result = uv_tcp_init(global_ev.loop, uv_tcp);
+    event_loop_unlock(&global_ev);
+
+    if (result != 0) {
+        free(uv_tcp);
+        free(tcp_socket);
+        std::string err = std::string("failed to initialize tcp_socket: ") + uv_strerror(result);
+        return io_result_mk_error(err.c_str());
+    }
+
+    tcp_socket->m_uv_tcp = uv_tcp;
+
+    lean_object * obj = lean_uv_tcp_socket_new(tcp_socket);
+    lean_mark_mt(obj);
+
+    tcp_socket->m_uv_tcp->data = obj;
+
+    return lean_io_result_mk_ok(obj);
+}
+
+lean_object* create_promise() {
+    lean_object * prom_res = lean_io_promise_new(lean_io_mk_world());
+    lean_object * promise = lean_ctor_get(prom_res, 0);
+    lean_inc(promise);
+    lean_dec(prom_res);
+    return promise;
+}
+
+void resolve_promise(lean_object* promise, int status) {
+    lean_object * res;
+
+    if (status == 0) {
+        res = alloc_cnstr(1, 1, 0);
+        cnstr_set(res, 0, lean_box(0));
+    } else {
+        res = mk_except_error_string(uv_strerror(status));
+    }
+
+    lean_object * result = lean_io_promise_resolve(res, promise, lean_io_mk_world());
+    lean_dec(result);
+}
+
+extern "C" LEAN_EXPORT lean_obj_res lean_uv_tcp_connect(b_obj_arg socket, b_obj_arg addr) {
+    lean_uv_tcp_socket_object * tcp_socket = lean_to_uv_tcp_socket(socket);
+
+    sockaddr_in addr_ptr;
```

Comment:
As this function is not specific to IPv4 this should probably not be a `sockaddr_in`?